### PR TITLE
Change default SSH port and introduce cloud-init support

### LINKED ISSUE #1613: DigitalOcean: Failed to connect to host
Deployment host: Ubuntu 18.04
Default config.cfg (users changed / added)
Amsterdam region in DO

Errors out with server unreachable.
Server is created in DO and is reachable over port 22

Log:
`<pre>.env) root@ubuntu:/home/thomas/algo# ./algo
<font color="#75507B"><b> [WARNING]: Could not match supplied host pattern, ignoring: vpn-host</b></font>

PLAY [localhost] **********************************************************************************************************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [Ensure the requirements installed] **********************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [Set required ansible version as a fact] *****************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost] =&gt; (item=ansible==2.8.3)</font>

TASK [Verify Python meets Algo VPN requirements] **************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost] =&gt; {</font>
<font color="#4E9A06">    &quot;changed&quot;: false,</font>
<font color="#4E9A06">    &quot;msg&quot;: &quot;All assertions passed&quot;</font>
<font color="#4E9A06">}</font>

TASK [Verify Ansible meets Algo VPN requirements] *************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost] =&gt; {</font>
<font color="#4E9A06">    &quot;changed&quot;: false,</font>
<font color="#4E9A06">    &quot;msg&quot;: &quot;All assertions passed&quot;</font>
<font color="#4E9A06">}</font>

PLAY [Ask user for the input] *********************************************************************************************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>
[Cloud prompt]
What provider would you like to use?
    1. DigitalOcean
    2. Amazon Lightsail
    3. Amazon EC2
    4. Microsoft Azure
    5. Google Compute Engine
    6. Hetzner Cloud
    7. Vultr
    8. Scaleway
    9. OpenStack (DreamCompute optimised)
    10. CloudStack (Exoscale optimised)
    11. Install to existing Ubuntu 18.04 or 19.04 server (Advanced)
  
Enter the number of your desired provider
:

TASK [Cloud prompt] *******************************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [Set facts based on the input] ***************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>
[VPN server name prompt]
Name the vpn server
[algo]
:

TASK [VPN server name prompt] *********************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>
[Cellular On Demand prompt]
Do you want macOS/iOS IPsec clients to enable &quot;Connect On Demand&quot; when connected to cellular networks?
[y/N]
:

TASK [Cellular On Demand prompt] ******************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>
[Wi-Fi On Demand prompt]
Do you want macOS/iOS IPsec clients to enable &quot;Connect On Demand&quot; when connected to Wi-Fi?
[y/N]
:

TASK [Wi-Fi On Demand prompt] *********************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>
[Retain the PKI prompt]
Do you want to retain the keys (PKI)? (required to add users in the future, but less secure)
[y/N]
:

TASK [Retain the PKI prompt] **********************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>
[DNS adblocking prompt]
Do you want to enable DNS ad blocking on this VPN server?
[y/N]
:

TASK [DNS adblocking prompt] **********************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>
[SSH tunneling prompt]
Do you want each user to have their own account for SSH tunneling?
[y/N]
:

TASK [SSH tunneling prompt] ***********************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [Set facts based on the input] ***************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

PLAY [Provision the server] ***********************************************************************************************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

--&gt; Please include the following block of text when reporting issues:

Algo running on: Ubuntu 19.04 (Virtualized: vmware)
Created from git clone. Last commit: 88eaf30 Update README.md (#1602)
Python 3.7.3
Runtime variables:
    algo_provider &quot;digitalocean&quot;
    algo_ondemand_cellular &quot;False&quot;
    algo_ondemand_wifi &quot;False&quot;
    algo_ondemand_wifi_exclude &quot;X251bGw=&quot;
    algo_dns_adblocking &quot;True&quot;
    algo_ssh_tunneling &quot;False&quot;
    wireguard_enabled &quot;True&quot;
    dns_encryption &quot;True&quot;

TASK [Display the invocation environment] *********************************************************************************************************************************************************************
<font color="#C4A000">changed: [localhost -&gt; localhost]</font>

TASK [Install the requirements] *******************************************************************************************************************************************************************************
<font color="#C4A000">changed: [localhost -&gt; localhost]</font>

TASK [Generate the SSH private key] ***************************************************************************************************************************************************************************
<font color="#C4A000">changed: [localhost]</font>

TASK [Generate the SSH public key] ****************************************************************************************************************************************************************************
<font color="#C4A000">changed: [localhost]</font>
[cloud-digitalocean : pause]
Enter your API token. The token must have read and write permissions (https://cloud.digitalocean.com/settings/api/tokens):
 (output is hidden):

TASK [cloud-digitalocean : pause] *****************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [cloud-digitalocean : Set the token as a fact] ***********************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [cloud-digitalocean : Get regions] ***********************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [cloud-digitalocean : Set facts about the regions] *******************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [cloud-digitalocean : Set default region] ****************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>
[cloud-digitalocean : pause]
What region should the server be located in?
    1. ams3     Amsterdam 3
    2. blr1     Bangalore 1
    3. fra1     Frankfurt 1
    4. lon1     London 1
    5. nyc1     New York 1
    6. nyc3     New York 3
    7. sfo2     San Francisco 2
    8. sgp1     Singapore 1
    9. tor1     Toronto 1
  
Enter the number of your desired region
[6]
:

TASK [cloud-digitalocean : pause] *****************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [cloud-digitalocean : Set additional facts] **************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [cloud-digitalocean : Upload the SSH key] ****************************************************************************************************************************************************************
<font color="#C4A000">changed: [localhost]</font>

TASK [cloud-digitalocean : Creating a droplet...] *************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [cloud-digitalocean : set_fact] **************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [Set subjectAltName as a fact] ***************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [Add the server to an inventory group] *******************************************************************************************************************************************************************
<font color="#C4A000">changed: [localhost]</font>

TASK [Additional variables for the server] ********************************************************************************************************************************************************************
<font color="#C4A000">changed: [localhost]</font>

TASK [Wait until SSH becomes ready...] ************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [Linux | set OS specific facts] **************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [Set config paths as facts] ******************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

TASK [Update config paths] ************************************************************************************************************************************************************************************
<font color="#C4A000">changed: [localhost]</font>

TASK [debug] **************************************************************************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost] =&gt; {</font>
<font color="#4E9A06">    &quot;IP_subject_alt_name&quot;: &quot;178.62.244.24&quot;</font>
<font color="#4E9A06">}</font>
Pausing for 20 seconds
(ctrl+C then &apos;C&apos; = continue early, ctrl+C then &apos;A&apos; = abort)

TASK [A short pause, in order to be sure the instance is ready] ***********************************************************************************************************************************************
<font color="#4E9A06">ok: [localhost]</font>

PLAY [Configure the server and install required software] *****************************************************************************************************************************************************

TASK [common : Check the system] ******************************************************************************************************************************************************************************
<font color="#CC0000"><b>fatal: [178.62.244.24]: UNREACHABLE! =&gt; {&quot;changed&quot;: false, &quot;msg&quot;: &quot;Failed to connect to the host via ssh: Warning: Permanently added &apos;178.62.244.24&apos; (ECDSA) to the list of known hosts.\r\nroot@178.62.244.24: Permission denied (publickey).&quot;, &quot;unreachable&quot;: true}</b></font>

PLAY RECAP ****************************************************************************************************************************************************************************************************
<font color="#CC0000">178.62.244.24</font>              : ok=0    changed=0    <font color="#CC0000"><b>unreachable=1   </b></font> failed=0    skipped=0    rescued=0    ignored=0   
<font color="#C4A000">localhost</font>                  : <font color="#4E9A06">ok=39  </font> <font color="#C4A000">changed=8   </font> unreachable=0    failed=0    <font color="#06989A">skipped=3   </font> rescued=0    ignored=0   </pre>`

### LINKED ISSUE #1635: Custom port for ssh access
**Is your feature request related to a problem? Please describe.**
Running anything on port 22 on my digitalOcean droplet results in mile long attempts to login via ssh from all over the world. This is kind of frustrating, since I'm sure 99% of those attempts are simple port scanning and dicitonary attacks and nothing to worry about, except for the miles and miles of log files. Moving the sshd to any other port than 22 is an instant fix for that problem.

**Describe the solution you'd like**
Let users choose to run sshd on a different port. 

**Describe alternatives you've considered**
I managed to move it manually by changing the port in the sshd_config and adding an accept rule with the new port to the iptables, but I worry that I broke algo's installer and won't be able to e.g. add/remove users from my local machine now (at least not without moving sshd ports and poking holes into the DO firewall again)

**Additional context**
I hope I didn't miss anything when searching the repo and googling a bit. All I found so far was a "nope, you can't configure that".

### PR DESCRIPTION
**Pull request by jackivanov**
## Description
I would like to bring support for cloud-init as the first step to set up a server and make some changes that Ansible is not capable of.

## Motivation and Context
The motivations are to eliminate issues like #1613, and bring more security by changing the default SSH port (#1635), the default SSH user, and removing all others hardcoded users by the cloud provider. Cloud-init also might be used to solve more problems and speed up the deployment process. 

For Lightsail and Vultr  it is only possible with a bash script, for all the others we can use the native cloud-config. 

~The changes are not compatible with Scaleway~ For scaleway the uri module is being used (see ansible/ansible/issues/65036) and I'm thinking of removing this provider from Algo as I've been continuously fighting with their services, producing new bugs. I'm not sure the effort is worth any further support.

Fixes #1613, Closes #1635

## How Has This Been Tested?
Deployed to every cloud provider, but more tests should be done.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by davidemyers** [comment]
I like the idea of having a consistent user account across cloud providers.

The `bash` script should probably make the `sudoers` file mode `0440` and the `.ssh` directory mode `0700` to match `cloud-init`.

I'm not sure why the `algo` user needs to be in group `sudo` since you're creating a custom `sudoers` file, but it would be handy if it was at least in group `adm` to be able to read logs and use `journalctl` without having to use `sudo`.

Here's a patch for the firewall documentation:

```patch
diff --git a/docs/firewalls.md b/docs/firewalls.md
index 8feb508..3b76fcc 100644
--- a/docs/firewalls.md
+++ b/docs/firewalls.md
@@ -24,7 +24,7 @@ Any external firewall must be configured to pass the following incoming ports ov
 
 Port | Protocol | Description | Related variables in `config.cfg`
 ---- | -------- | ----------- | ---------------------------------
-22    | TCP | Secure Shell (SSH) | None
+4160  | TCP | Secure Shell (SSH) | `ssh_port`
 500   | UDP | IPsec IKEv2 | `ipsec_enabled`
 4500  | UDP | IPsec NAT-T | `ipsec_enabled`
 51820 | UDP | WireGuard | `wireguard_enabled`, `wireguard_port`
```

**Comment by jackivanov** [comment]
I think this is ready for tests and reviews

File: tests/cloud-init.sh [hunk 1]
```diff
@@ -9,7 +9,7 @@ export DNS_ADBLOCKING=true
 export SSH_TUNNELING=true
 export ENDPOINT=10.0.8.100
 export USERS=desktop,user1,user2
-export EXTRA_VARS='install_headers=false tests=true apparmor_enabled=false local_service_ip=172.16.0.1'
+export EXTRA_VARS='install_headers=false tests=true apparmor_enabled=false local_service_ip=172.16.0.1 ansible_ssh_user=4160'
```
Comment (davidemyers) [inline tests/cloud-init.sh]: `ansible_ssh_user=4160` doesn't look right.

File: config.cfg [hunk 2]
```diff
@@ -134,7 +138,7 @@ congrats:
   ca_key_pass: |
     "#        The CA key password is {{ CA_password|default(omit) }}       #"
   ssh_access: |
-    "#      Shell access: ssh -i {{ ansible_ssh_private_key_file|default(omit) }} {{ ansible_ssh_user|default(omit) }}@{{ ansible_ssh_host|default(omit) }}        #"
+    "#      Shell access: ssh -i {{ SSH_keys.private }} {{ ansible_ssh_user|default(omit) }}@{{ ansible_ssh_host|default(omit) }} -p {{ ssh_port }}        #"
```
Comment (davidemyers) [inline config.cfg]: Should this be the `-F` form of the command?

Comment (jackivanov) [reply config.cfg]: Fixed, thanks!

File: tests/cloud-init.sh (hunk 1 above)
Comment (jackivanov) [reply tests/cloud-init.sh]: Fixed, thanks

**Pull request merged**